Move NS auxkernels away from AD

GiudGiud opened the pull request:
refs https://github.com/idaholab/moose/issues/31195

most of the work was already done in the linear K-epsilon PR

GiudGiud committed:
Transition auxkernels to non-AD in the NS module
refs #31195

GiudGiud committed:
Template NS utilities to be able to use them without AD too
refs #31195

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3092758/), step Docs: sync website on b150d13 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31196/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3092773/), step Generate coverage on b150d13 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3cca8</th>
      <th colspan=4>#31196 b150d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.01%</td>
      <td>86.00%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8180</td>
      <td>8179</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1331</td>
      <td>1332</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31196/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31196/coverage/fluid_properties)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3cca8</th>
      <th colspan=4>#31196 b150d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.39%</td>
      <td>85.39%</td>
      <td>-0.01%</td>
      <td>94.64%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>19760</td>
      <td>19748</td>
      <td>-12</td>
      <td>53</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3380</td>
      <td>3380</td>
      <td>-</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31196/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31196/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31196/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionWallShearStressAux.C:
```diff
@@ -98,31 +98,30 @@ WallFunctionWallShearStressAux::computeValue()
 
   // Get the velocity vector
-  ADRealVectorValue velocity(_u_var->getElemValue(&elem, state));
+  RealVectorValue velocity(_u_var->getElemValue(&elem, state).value());
```
```suggestion
  RealVectorValue velocity(MetaPhysicL::raw_value(_u_var->getElemValue(&elem, state)));
```
Let's try to be consistent

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionWallShearStressAux.C:
```diff
@@ -98,31 +98,30 @@ WallFunctionWallShearStressAux::computeValue()
+  RealVectorValue velocity(_u_var->getElemValue(&elem, state).value());
   if (_v_var)
-    velocity(1) = _v_var->getElemValue(&elem, state);
+    velocity(1) = _v_var->getElemValue(&elem, state).value();
```
```suggestion
    velocity(1) = MetaPhysicL::raw_value(_v_var->getElemValue(&elem, state));
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionWallShearStressAux.C:
```diff
@@ -98,31 +98,30 @@ WallFunctionWallShearStressAux::computeValue()
+    velocity(1) = _v_var->getElemValue(&elem, state).value();
   if (_w_var)
-    velocity(2) = _w_var->getElemValue(&elem, state);
+    velocity(2) = _w_var->getElemValue(&elem, state).value();
```
```suggestion
    velocity(2) = MetaPhysicL::raw_value(_w_var->getElemValue(&elem, state));
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionWallShearStressAux.C:
```diff
@@ -98,31 +98,30 @@ WallFunctionWallShearStressAux::computeValue()
   const auto mu = _mu(elem_arg, state);
-  ADReal u_star = NS::findUStar<ADReal>(mu.value(), rho.value(), parallel_speed, dist);
-  ADReal tau = u_star * u_star * rho.value();
+  Real u_star = NS::findUStar<Real>(mu, rho, parallel_speed, dist);
+  Real tau = u_star * u_star * rho;
```
```suggestion
  const Real u_star = NS::findUStar<Real>(mu, rho, parallel_speed, dist);
  const Real tau = u_star * u_star * rho;
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionYPlusAux.C:
```diff
@@ -98,31 +98,29 @@ WallFunctionYPlusAux::computeValue()
 
   // Get the velocity vector
-  ADRealVectorValue velocity(_u_var->getElemValue(&elem, state));
+  RealVectorValue velocity(_u_var->getElemValue(&elem, state).value());
```
```suggestion
  RealVectorValue velocity(MetaPhysicL::raw_value(_u_var->getElemValue(&elem, state)));
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionYPlusAux.C:
```diff
@@ -98,31 +98,29 @@ WallFunctionYPlusAux::computeValue()
+    velocity(1) = _v_var->getElemValue(&elem, state).value();
   if (_w_var)
-    velocity(2) = _w_var->getElemValue(&elem, state);
+    velocity(2) = _w_var->getElemValue(&elem, state).value();
```
```suggestion
    velocity(2) = MetaPhysicL::raw_value(_w_var->getElemValue(&elem, state));
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionYPlusAux.C:
```diff
@@ -98,31 +98,29 @@ WallFunctionYPlusAux::computeValue()
-  ADRealVectorValue parallel_velocity = velocity - perpendicular_speed * normal;
-  ADReal parallel_speed = parallel_velocity.norm();
-  ADRealVectorValue parallel_dir = parallel_velocity / parallel_speed;
+  Real perpendicular_speed = velocity * normal;
+  RealVectorValue parallel_velocity = velocity - perpendicular_speed * normal;
+  Real parallel_speed = parallel_velocity.norm();
```
```suggestion
  const Real perpendicular_speed = velocity * normal;
  const RealVectorValue parallel_velocity = velocity - perpendicular_speed * normal;
  const Real parallel_speed = parallel_velocity.norm();
```

lindsayad commented on modules/navier_stokes/src/auxkernels/WallFunctionYPlusAux.C:
```diff
@@ -98,31 +98,29 @@ WallFunctionYPlusAux::computeValue()
   const auto mu = _mu(elem_arg, state);
-  ADReal u_star = NS::findUStar<ADReal>(mu.value(), rho.value(), parallel_speed, dist);
-  ADReal tau = u_star * u_star * rho;
+  Real u_star = NS::findUStar<Real>(mu, rho, parallel_speed, dist);
```
```suggestion
  const Real u_star = NS::findUStar<Real>(mu, rho, parallel_speed, dist);
```

lindsayad commented on modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:
```diff
@@ -133,7 +133,7 @@ INSFVMomentumAdvection::computeResidualsAndAData(const FaceInfo & fi)
                                            &limiter_time};
     if (const auto [is_jump, eps_elem_face, eps_neighbor_face] =
-            NS::isPorosityJumpFace(epsilon(), fi, state);
+            NS::isPorosityJumpFace<ADReal>(epsilon(), fi, state);
```
```suggestion
            NS::isPorosityJumpFace(epsilon(), fi, state);
```

GiudGiud replied:
for these 3 calls the auto-detection is not working.

it gets hung up on the conversion epsilon, a FunctorBase, to a FunctorEnvelope (`Functor<T>`)

lindsayad replied:
Let's change `isPorosityJumpFace` to take `FunctorBase` then. Then we don't have an implicit conversion happening

lindsayad commented on modules/navier_stokes/src/variables/BernoulliPressureVariable.C:
```diff
@@ -135,7 +135,7 @@ BernoulliPressureVariable::isDirichletBoundaryFace(const FaceInfo & fi,
       return false;
 
-    if (std::get<0>(NS::isPorosityJumpFace(*_eps, fi, time)))
+    if (std::get<0>(NS::isPorosityJumpFace<ADReal>(*_eps, fi, time)))
```
```suggestion
    if (std::get<0>(NS::isPorosityJumpFace(*_eps, fi, time)))
```

GiudGiud committed:
Address Alex's review

- more const of local variables
- use template parameter auto-detection
- use raw_value()

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
all accepted

GiudGiud committed:
Use FunctorBase in porosity jump util
Allows all users to do auto-detection of type
Per Alex's suggestion

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3119984/), step Clang format on eb38734 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31196/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31196/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 419b9d0028143f1f75e636d542d17e2f93f452ea`

lindsayad reviewed (commented):
Did you accidentally drop a commit? You resolved a lot of my suggestions without incorporating them

GiudGiud commented:
must be what happened

GiudGiud commented:
it's back

GiudGiud commented on modules/navier_stokes/src/auxkernels/INSFVMixingLengthTurbulentViscosityAux.C:
```diff
@@ -58,16 +58,16 @@ INSFVMixingLengthTurbulentViscosityAux::computeValue()
 
-  const auto & grad_u = _u_var->adGradSln(&elem, state);
-  ADReal symmetric_strain_tensor_norm = 2.0 * Utility::pow<2>(grad_u(0));
+  const auto & grad_u = MetaPhysicL::raw_value(_u_var->adGradSln(&elem, state));
```
funny thing is that I did look at this line to check after force pushing why I needed to force push, read it backwards, and was going to write "fairly sure we dont need to take this by reference" but then abstained and moved on to something else
